### PR TITLE
fetch conditional GET with ETags

### PR DESCRIPTION
**Pull request by FND**
As discussed with @Ms2ger on IRC (FreeNode #whatwg), I've noticed that `fetch` doesn't seem to support conditional GETs by default.

I'm not sure whether that's by design, but @annevk et al. suggested it might be a bug.

This test case is intended to demonstrate that behavior.

Here's what I expected:

```
> GET /foo

< 200 OK
< ETag: "foo"
```

```
> GET /foo
> If-None-Match: "foo"

< 304 Not Modified
```

However, that second request does not include an `If-None-Match` header. (Tested with Firefox 48.0.1 and Chrome 53.0.2785.116 on OS X.)

I've created a new directory for this test case because I didn't know where else it might fit. Feel free to point out whether I'm violating any code conventions or suggest other adjustments.

---

Just to confirm that the server works as expected:

```
$ curl -v http://web-platform.test:8000/fetch/bugs/resources/cache.py; echo
> GET /fetch/bugs/resources/cache.py HTTP/1.1
> Host: web-platform.test:8000
> User-Agent: curl/7.43.0
> Accept: */*
>
< HTTP/1.1 200 OK
< ETag: "123abc"
< Content-Type: text/plain
< Server: BaseHTTP/0.3 Python/2.7.10
< Date: Fri, 26 Aug 2016 15:29:10 GMT
< Content-Length: 26
<
lorem ipsum dolor sit amet
```

```
$ curl -v -H 'If-None-Match: "123abc"' http://web-platform.test:8000/fetch/bugs/resources/cache.py
> GET /fetch/bugs/resources/cache.py HTTP/1.1
> Host: web-platform.test:8000
> User-Agent: curl/7.43.0
> Accept: */*
> If-None-Match: "123abc"
>
< HTTP/1.1 304 Not Modified
< Server: BaseHTTP/0.3 Python/2.7.10
< Date: Fri, 26 Aug 2016 15:29:17 GMT
< Content-Length: 0
<
```

### CODE REVIEW

File: fetch/bugs/request-etag.html [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+    <meta charset="utf-8">
+    <title>Request ETag</title>
+    <meta name="help" href="https://fetch.spec.whatwg.org/#request">
+    <script src="/resources/testharness.js"></script>
+    <script src="/resources/testharnessreport.js"></script>
+  </head>
+  <body>
+    <script>
+    promise_test(function() {
+      var URL = "resources/cache.py";
```
Comment (annevk) [inline fetch/bugs/request-etag.html]: URL is also on the global, might be better to use a different variable name here.

Comment (FND) [inline fetch/bugs/request-etag.html]: Good catch - I couldn't discern any particular pattern here; would `ENDPOINT` be acceptable?

**Comment by annevk** [comment]
So the main problem with this test is that it only works once. Maybe an idea is to generate a random token. And then use that for the first and second request in the URL somehow. And let the server use that random token as part of the ETag header. That way when you run the test again you'll get a new random token and no cache hits.

File: fetch/bugs/request-etag.html (hunk 1 above)
Comment (annevk) [inline fetch/bugs/request-etag.html]: Yeah or lowercase `url`. Doesn't matter much.

**Comment by annevk** [comment]
I guess I'd put the test in fetch/api/request somewhere. It's somewhat similar to fetch/api/request/request-cache.html actually. Maybe this is already tested there?

**Comment by FND** [comment]
> the main problem with this test is that it only works once

Perhaps I'm misunderstanding you, but it's not supposed to work at all: This test case is supposed to illustrate what seems like a bug in browser implementations (see description above), so it's should be failing:

```javascript
// runner-results.json
{
  "results": [
    {
      "test": "/fetch/bugs/request-etag.html",
      "subtests": [
        {
          "name": "Testing ETag handling",
          "status": "FAIL",
          "message": "assert_equals: expected 304 but got 200"
        }
      ],
      "status": "OK",
      "message": null
    }
  ]
}
```
(just confirmed again; I've updated OP's browser versions accordingly)

> It's somewhat similar to fetch/api/request/request-cache.html actually. Maybe this is already tested there?

Unfortunately, I couldn't definitively decipher what's going on there (too much abstraction). But while it does seem to test something similar, it didn't exhibit the issue I was observing (see above), so there's some sort of mismatch here.

**Comment by youennf** [comment]
In some test environments, we might want to rerun the test several times and the test should behave the same on every run.

In your test case, depending on the cache state, the first fetch might serve a cached response.
One possible fix is to use a dynamic request URL:  instead of "resources/cache.py", you could use "resources/cache.py?" + token() for instance.

**Comment by FND** [comment]
Thanks for the clarification. While that does make sense, the reason I remain puzzled is that I'm never even getting to the point where caching kicks in - which is why I ran into this issue in the first place. So am I correct in assuming that this test passes for you two, meaning I'm observing different behavior on my end?

Either way, I've added a cache-busting parameter in 814c7ac8b76bfbc06a97af798d99d9f06da31eea.

**Comment by youennf** [comment]
I think the comment is more general after checking the test code.
I haven't yet run the test but will do it.

**Comment by youennf** [comment]
As per https://fetch.spec.whatwg.org/#http-network-or-cache-fetch, a 304 response received at the network layer is converted as a 200 response using the browser cache at JS level.

Your test is expecting a 304 response at JS level but it will never surface.
What you really want is to check the 304 response at the network level.
You can check that using a web inspector or other network sniffing tools.

**Comment by youennf** [comment]
Just tried and chrome/firefox inspectors show that a 304 response is received.

**Comment by annevk** [comment]
Yeah, under certain conditions (defined in Fetch) we do want 304 to appear at the JS level. That allows someone to maintain their own HTTP cache implementation, e.g., in Indexed DB.

**Comment by youennf** [comment]
Interesting.
If the script adds an If-None-Match header to the second request, the 304 response should show up to JS as per the spec...

It would be great to update the test accordingly.

For consistency with other WPT tests, I would also use token() instead of Math.random()
The html test just needs to add <script src="/common/utils.js"></script>

**Comment by FND** [comment]
> a 304 response received at the network layer is converted as a 200 response using the browser cache at JS level

Well, that certainly explains it - so it's not a bug after all, just something I hadn't considered (in my defense, it's rather unintuitive). Thanks for clarifying that, and sorry for the noise.

I've updated the test accordingly, including @youennf's suggestions.

**Comment by youennf** [comment]
Should this test be merged?

**Comment by annevk** [comment]
@youennf yeah I think so, you don't have those rights or are you asking for a different reason?

**Comment by youennf** [comment]
I don't think I have those rights

**Comment by annevk** [comment]
I pinged someone on IRC to give you more rights, it seems like you should be able to push. Merged this one myself.

@FND thanks a lot for this contribution, hope more will come when you're bitten by other bugs 😊

**Pull request merged**